contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,103 @@
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+## Implementation details
+
+### API changes
+
+```go
+type Service struct {
+  spec:
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // TopologyKeyList is a topologyKey list, which is used to guarantee ability. For example,
+    // `TopologyKeyList: [ "kubernetes.io/hostname", "failure-domain.beta.kubernetes.io/zone" ]`
+    // If failed to match hostname, then it will try to find the endpoints in the same zone
+    // The priority of key in TopologyKeyList is sorted by index and selecting process will stop once a key matches.
+  	TopologyKeyList []string
+}
+```
+
+Although `NodeName` was already added to `EndpointAddress`, we want `Endpoints` to carry more node's topology informations so that allowing more topology levels other than hostname. So, create some annotations for Endpoints which is used toidentify what topology domain the endpoints pod exists, e.g. what host, rack,zone, region etc. In other words, the key of the annotations is the topology key and the value of the annotations is the corresponding value of node's labels.
+
+## Endpoints Controller changes
+
+Endpoint Controller will fill in the `Annotations` for `Endpoints`. We want `Endpoints.Annotations` to tell kube-proxy what topology domain(e.g. host, rack, zone, region etc.) the endpoints is in.
+
+Endpoints controller will need to watch ALL nodes and maintain a key-value store for node cache - the key is the nodename and value is the node object. So, the new logic of endpoint controller might like:
+
+```go
+go wath service, pod, node
+for topologyKey in service.TopologyKeyList; do
+  for pod in service backends; do
+    node := nodeCache[pod.Spec.NodeName]
+    // copy node labels about topology keys to endpoints.Annotations
```

Comment:
I don't think we can know what is "about" topology.  I think we have to carry all such.

Also, I think we want labels, not annotations.

Reply:
> I don't think we can know what is "about" topology. I think we have to carry all such.

Agree, you are right. After all, topology is arbitrary.

Reply:
> Also, I think we want labels, not annotations.

Yeah, labels seems better than annotations because we are copying node labels.